### DISCUSSION TITLE
Android Studio support

### DISCUSSION THREAD

**Discussion by Chevron4** [ani-memes/AMII#104] [Ideas]
Hi there! :) 
First of all, let me say that I'm a huge fan of your plugin. You guys saved me from countless hours of frustration by showing me awesome gifs after many failed build attempts. A big thank you to you guys for that, please keep up the good work!
Now to my request:
Do you guys have any plans to support Android Studio events? I use the plugin with several projects of mine in IntelliJ Ultimate and it works very well. However, when I work on apps with Android Studio / its integration in  IntelliJ Ultimate, it ignores the build events...

**Comment by Unthrottled** [comment]
>  when I work on apps with Android Studio / its integration in IntelliJ Ultimate

So are you doing Android work in `Android Studio` or `IntelliJ Idea Ultimate`?

> it ignores the build events...

What kind of build events? Is there another type of build? Also are you expecting reactions when the build fails or passes?

Also it's nice to hear you are enjoying my plugin :)

**Reply by Chevron4** [reply]
Thank you for your fast response! 
In other projects in IntelliJ Ultimate, gifs are shown to me after build, whether successful or not. When I build Android apps with Android Studio or IntelliJ Ultimate, the gifs don't work properly.

IntelliJ Ultimate: If the "Build project" button (hammer icon) is pressed, then the plugin responds and shows a gif. If the "Run" button (Play icon) is pressed, then the plugin does not respond.

Android Studio: The plugin does not react to the "Make project" button (hammer icon), nor to the "Run" button (play icon).

In both programs the feature "watch logs" does not work either. Here I tested different cases like "Log.d("TAG", "MESSAGE") / Log.d("TAG", "MESSAGE") / Log.w("TAG", "MESSAGE")" for the Logcat or Gradle build logs like "println("MESSAGE") / logger.debug("MESSAGE") / logger.warn("MESSAGE")".

Startup and idle gifs work without problems.

The behavior can be recreated in both programs by simply creating a new "Basic Activity" project using Kotlin.

The versions I used:
Anime Memes 0.9.3

Android Studio 4.2
Build #AI-202.7660.26.42.7322048, built on April 29, 2021

IntelliJ IDEA 2021.1.1 (Ultimate Edition)
Build #IU-211.7142.45, built on April 30, 2021

Sorry for my previous vague description :)

**Reply by Unthrottled** [reply]
> In other projects in IntelliJ Ultimate, gifs are shown to me after build, whether successful or not.

That's not how the feature should be working. So you'll get a negative notification only if you smash the hammer and the build fails. A positive reaction only occurs when you smash the hammer and it passes and the previous build failed. This prevents the the plugin from being too chatty. You can make a feature request for adding a non-default config that enables reactions on all builds though.

> IntelliJ Ultimate: If the "Build project" button (hammer icon) is pressed, then the plugin responds and shows a gif. If the "Run" button (Play icon) is pressed, then the plugin does not respond.

`If the "Run" button is pressed...` unless the running causes the plugin to build the project, the only time you should receive a reaction if the process exits with a non-zero code. (It looks like android studio doesn't exit with codes at first glance.)

> Android Studio: The plugin does not react to the "Make project" button (hammer icon)...

I've verified this and it is a bug. The fix will probably take the shape of a new extension that leverages the Android studio build APIs. Because apparently they didn't want to use the common platform ones *shakes fist at android studio*.

**Reply by Chevron4** [reply]
> That's not how the feature should be working.

I was already aware of this by using the plugin with non-Android projects, I think the current behavior is great as it is! :) With that, I just wanted to say that the plugin reacts to the build. (I have lowered the value of "events before frustration" for testing.)

> (It looks like android studio doesn't exit with codes at first glance.)

Ah okay, I was not aware of that, sorry.

> Because apparently they didn't want to use the common platform ones shakes fist at android studio.

Oh dear...that would be too easy...then maybe something would work right away -.-' :D

**Reply by Unthrottled** [reply]
> In both programs the feature "watch logs" does not work either. Here I tested different cases like "Log.d("TAG", "MESSAGE") / Log.d("TAG", "MESSAGE") / Log.w("TAG", "MESSAGE")" for the Logcat or Gradle build logs like "println("MESSAGE") / logger.debug("MESSAGE") / logger.warn("MESSAGE")".

:point_up: That's currently a bug in the plugin itself, I accidentally broke it, it should be fixed when #109 is fixed.

Will let you know when the build event stuff works though.

**Reply by Chevron4** [reply]
Thank you very much! :)

**Reply by Unthrottled** [reply]
Alright, I found android's build process API.

I had to create another extension that is dependent on that API and AMII, that way only people with the Android plugin need to add this and not everybody else.

It isn't published to the marketplace yet, so you'll have to follow the `Manual` install process.
[Here is the Android extension](https://github.com/ani-memes/amii-android-extension/tree/main#installation)

Thanks for bringing this to my attention :clap:

**Reply by Chevron4** [reply]
I've been using it since your upload and it works just fantastic. Thanks again! :)

**Reply by Unthrottled** [reply]
Awesome! Thanks for the follow up :smiley:
